circuits overview & linkedmultiquery additions (#235)
0xPolygonID/docs: docs/verifier/circuits.md

---
id: circuits
title: "Circuits"
sidebar_label: "Circuits"
description: "Explore the essential Circom circuits used in Iden3’s decentralized identity protocol. Learn how state transitions, authentication, and advanced credential verification techniques work together for robust digital identity management."
keywords:
  - iden3
  - circuits
  - decentralized identity
  - identity verification
  - circom
  - zkSNARKs
  - state transition
  - authentication
  - credential verification
  - merkle tree
  - on-chain
  - off-chain
  - digital identity
  - cryptography
---

The following is the overview of the most important `circom` circuits used in Privado ID protocol and ecosystem. These circuits play a crucial role in enabling secure identity ownership, trustworthy credential verification, and efficient on-chain or off-chain checks.
## 1. stateTransition

**Circuit Purpose:**  

> Validates an identity state transition from an `oldUserState` to a `newUserState`, ensuring the identity owner is the one making the update.

Reviewer: ```suggestion
Validates an identity state transition from an old user state to the new one, ensuring the identity owner is the one making the update.
```

## 3. credentialAtomicQueryMTPV2

**Circuit Purpose:**  
Proves that the Prover holds a valid credential (issued by an issuer) that satisfies a Verifier’s query. The credential's representation is stored in the issuer’s Claims Tree, and the circuit relies on:

- Checking that the core claim representation hash is included in this tree via a Merkle Tree Proof (MTP) approach.
- Verifying the non-revocation of the credentials in the issuer’s Revocation Tree.
- Checking the validity of the issuer's state and its existence in the state contract.
- Ensuring that the Prover uses an identity that controls the credential subject identifier.
- Supporting queries on any credential field, thanks to the unique JSON-LD merklization algorithm.

**Usage Notes:**

- Designed for off-chain verifications with atomic queries.

## 4. credentialAtomicQueryMTPV2Onchain

**Circuit Purpose:**  
Same as `credentialAtomicQueryMTPV2`, but with support for output optimizations, making it more efficient for on-chain use. **Note:** It does not support selective disclosure.

**Usage Notes:**

- Designed for on-chain verifiers.

> - Relies on a trusted setup (typical of v2 circuits).

Reviewer: ```suggestion
```

## 5. credentialAtomicQuerySigV2

**Circuit Purpose:**  
Similar to `credentialAtomicQueryMTPV2` but for credentials that are signed by the issuer's BabyJubJub key rather than being included in the issuer’s Merkle Tree. It:

- Verifies the issuer’s signature over the core claim representation of a W3C credential (including the Merkle root of the JSON-LD document).

> - Checks that the BabyJubJub key is included in the issuer state (which is either at genesis or published on-chain).

Reviewer: ```suggestion
- Checks that the BabyJubJub key is included in the issuer state (which is either at genesis or published on-chain). and that key is not revoked in the latest issuer state.
```

## 8. credentialAtomicQueryV3Onchain-beta.1

**Circuit Purpose:**  
Designed for smart contract verifiers, this circuit builds on `credentialAtomicQueryV3` by adding two major enhancements:

1. It can check that the Prover controls the identity, similar to how `authV2` works (if authentication is enabled).
2. It computes a hash of all query-related inputs (e.g., `claimSchema`, `slotIndex`, `operator`, `claimPathKey`, `claimPathNotExists`, and any field values) so they don’t need to be exposed individually as public inputs on-chain.

**Usage Notes:**

- Ideal for on-chain verification where minimizing public inputs is crucial to reduce gas costs.
- Retains all capabilities of `credentialAtomicQueryV3`, such as selective disclosure, expiration checks, and optional revocation checks.
- If authentication is enabled, the Prover must demonstrate identity ownership in the same way as `authV2` before the proof is accepted.
- The reduced set of public inputs lowers the overhead for smart contract verification, making it more efficient for complex or frequent credential proofs on Ethereum.

> ## 9. linkedMultiQuery10.circom

Reviewer: ```suggestion
## 9. linkedMultiQuery10-beta.1
```
